Source/Core/InputCommon/ControllerInterface/GameController/GameController.mm

```diff
@@ -0,0 +1,80 @@
+// Copyright 2013 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include <random>
+
+#import <objc/message.h>
+
+#include "InputCommon/ControllerInterface/GameController/GameController.h"
+#include "InputCommon/ControllerInterface/GameController/GameControllerDevice.h"
+
+#include "Common/Logging/Log.h"
+
+namespace ciface::GameController
+{
+static char *s_observer="";
+
+void addController(GCController* controller)
+{
+  controller.motion.sensorsActive = true;
+
+  std::string vendorName = std::string([controller.vendorName UTF8String]);
+  INFO_LOG_FMT(SERIALINTERFACE, "Controller '{}' connected with motion={}", vendorName, controller.motion.sensorsActive);
+  NOTICE_LOG_FMT(SERIALINTERFACE, "Controller '{}' connected with motion={} and id=", vendorName, controller.motion.sensorsActive);
+
+  std::shared_ptr<ciface::GameController::Controller> j=std::make_shared<Controller>(controller);
+    
+  g_controller_interface.AddDevice(j);
+}
+
+static void onControllerConnect(CFNotificationCenterRef center, void* observer, CFStringRef name, const void* object, CFDictionaryRef userInfo)
+{
+  GCController* controller = (GCController*)object;
+  addController(controller);
+}
+
+static void onControllerDisconnect(CFNotificationCenterRef center, void* observer, CFStringRef name, const void* object, CFDictionaryRef userInfo)
+{
+  GCController *inController = (GCController*)object;
+    
+  std::string vendorName = std::string([inController.vendorName UTF8String]);
+  INFO_LOG_FMT(SERIALINTERFACE, "Controller '{}' disconnected", vendorName);
+    
+  g_controller_interface.RemoveDevice([&inController](const auto* obj)
+  {
+    const Controller* controller = dynamic_cast<const Controller*>(obj);
+    if (controller && controller->IsSameDevice(inController))
+      return true;
+
+    return false;
+  });
+}
+
+void Init(void* window)
+{
+  CFNotificationCenterAddObserver(CFNotificationCenterGetLocalCenter(), s_observer, &onControllerConnect, CFSTR("GCControllerDidConnectNotification"), NULL, CFNotificationSuspensionBehaviorDeliverImmediately);
+
+  CFNotificationCenterAddObserver(CFNotificationCenterGetLocalCenter(), s_observer, &onControllerDisconnect, CFSTR("GCControllerDidDisconnectNotification"), NULL, CFNotificationSuspensionBehaviorDeliverImmediately);
+}
+
+void PopulateDevices(void* window)
+{
+  // I was unable to use [GCController controllers] in Objective-C++ and therefore needed to access this class method using objc_msgSend instead
+  typedef NSArray<GCController*>* (*send_type)(id, SEL);
+  send_type func = (send_type)objc_msgSend;
+  NSArray<GCController*>* controllers=(NSArray<GCController*>*)func(objc_getClass("GCController"), sel_getUid("controllers"));
+
+  for (GCController* controller in controllers)
+  {
+    addController(controller);
```

Comment:
You should remove all devices which have the same source name as ours before re-adding them.

Reply:
@Filoppi during my tests, it seems like this is working properly. When I click 'Refresh' in the Device popup screen the controllers are added properly again.

Reply:
It does work properly in the way the controller interface is currently setup, meaning that it cleans all devices before calling PopulateDevices() on each input backend, but it doesn't hurt to also remove your devices manually, and some input backends do. It would avoid duplicates if the controller interface design ever changed.